keps/sig-node/4540-strict-cpu-reservation/README.md

```diff
@@ -0,0 +1,465 @@
+
+To maintain backward compatibility, we introduce `numMinSharedCPUs` in KubeletConfiguration as the minimum number of CPU cores not available for exclusive allocation and expose it to Kube-scheduler.
+
+![MinSharedCPUs](./strict-cpu-allocation.png)
+
+When `strict-cpu-reservation` is disabled:
+```console
+Total CPU cores: 64
+ReservedSystemCPUs: 6
+defaultCPUSet = Reserved (6) + 58 (available for exclusive allocation)
+```
+
+When `strict-cpu-reservation` is enabled:
+```console
+Total CPU cores: 64
+ReservedSystemCPUs: 6
+MinSharedCPUs: 4
+defaultCPUSet = MinSharedCPUs (4) + 54 (available for exclusive allocation)
+```
+
+## Design Details
+
+In Kubelet, when `strict-cpu-reservation` is enabled as a policy option, we remove the reserved cores from the shared pool at the stage of calculation DefaultCPUSet and remove the `MinSharedCPUs` from the list of available cores for exclusive allocation.
+
+Feature impact can be illustrated as following:
+
+With the following Kubelet configuration:
+
+```yaml
+kind: KubeletConfiguration
+apiVersion: kubelet.config.k8s.io/v1beta1
+featureGates:
+  ...
+  CPUManagerPolicyOptions: true
+  CPUManagerPolicyAlphaOptions: true
+cpuManagerPolicy: static
+cpuManagerPolicyOptions:
+  strict-cpu-reservation: "true"
+reservedSystemCPUs: "0,32,1,33,16,48"
+numMinSharedCPUs: 4
+...
+```
+
+When `strict-cpu-reservation` is disabled:
+```console
+# cat /var/lib/kubelet/cpu\_manager\_state
+{"policyName":"static","defaultCpuSet":"0-79","checksum":1241370203}
+```
+
+When `strict-cpu-reservation` is enabled:
+```console
+# cat /var/lib/kubelet/cpu\_manager\_state
+{"policyName":"static","defaultCpuSet":"2-3,6-15,17-31,34-35,38-47,49-63","entries":{"6dda0e2e-ac6c-4f9d-8cff-201e18aebb2f":{"busybox":"36"},"6ed17b91-0eb0-4a0c-8841-3565fd2e45dc":{"busybox":"37"},"716e4823-ca49-4be7-a429-58d6f3c29d96":{"busybox":"4"},"93ad7328-0dd5-4f9c-ac51-eb46debb16a7":{"busybox":"5"}},"checksum":2153668133}
+```
+
+In Node API, we add `exclusive-cpu` in Node Allocatable for Kube-scheduler to consume.
+
+```
+  "status": {
+    "capacity": {
+      "cpu": "64",
+      "exclusive-cpu": "64",
+      "ephemeral-storage": "832821572Ki",
+      "hugepages-1Gi": "0",
+      "hugepages-2Mi": "0",
+      "memory": "196146004Ki",
+      "pods": "110"
+    },
+    "allocatable": {
+      "cpu": "58",
+      "exclusive-cpu": "54",
+      "ephemeral-storage": "767528359485",
+      "hugepages-1Gi": "0",
+      "hugepages-2Mi": "0",
+      "memory": "186067796Ki",
+      "pods": "110"
+    },
+  ...
+```
+
+In kube-scheduler, `ExlusiveMilliCPU` is added in scheduler's `Resource` structure and `NodeResourcesFit` plugin is extended to filter out nodes that can not meet pod's exclusive CPU request.
+
+A new item `ExclusiveMilliCPU` is added in the scheduler `Resource` structure:
+
+```
+// Resource is a collection of compute resource.
+type Resource struct {
+        MilliCPU          int64
+        ExclusiveMilliCPU int64    // added
```

Comment:
can a container ask for both `MilliCPU` and `Exclusive[Milli]CPU`? what's the semantic in this case?

Reply:
MilliCPU is the total, ExclusiveMilliCPU is the exclusive portion, same semantics as for Node Allocatable.